### PR TITLE
Bring XmlReader{Async} to one behavior

### LINKED ISSUE #52646: XmlReader.ReadContentAs{Async} different behavior
There is a difference in reading content depending on the settings XmlReaderSettings
```csharp
using System;
using System.IO;
using System.Xml;
using System.Xml.Schema;

const string xml = "<test a=\"sdf\"/>";
//1
try
{
	using var xmlReader = XmlReader.Create(new StringReader(xml));
	xmlReader.MoveToContent();
	xmlReader.MoveToAttribute(0);

	if (xmlReader.ReadContentAs(typeof(XmlAtomicValue), null) is XmlAtomicValue xmlAtomic)
	{
//{"Content cannot be converted to the type System.Xml.Schema.XmlAtomicValue. Line 1, position 7."}
		Console.WriteLine(xmlAtomic.Value);
	}
}
catch (Exception ex)
{
	Console.WriteLine(ex);
}
//2
try
{
	using var xmlReader = XmlReader.Create(new StringReader(xml), new XmlReaderSettings {Async = true});
	await xmlReader.MoveToContentAsync();
	xmlReader.MoveToAttribute(0);

	if (await xmlReader.ReadContentAsAsync(typeof(XmlAtomicValue), null) is XmlAtomicValue xmlAtomic)
	{
//ok
		Console.WriteLine(xmlAtomic.Value);
	}
}
catch (Exception ex)
{
	Console.WriteLine(ex);
}
```

### PR DESCRIPTION
**Pull request by kronic**
see
[XmlRederAsync.ReadElementContentAsAsync](https://github.com/dotnet/runtime/blob/aa2af309a12d31dd68de9dc3646d77b6f25d9d40/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs#L106)
[XmlRederAsync.ReadContentAsAsync](https://github.com/dotnet/runtime/blob/aa2af309a12d31dd68de9dc3646d77b6f25d9d40/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs#L61)
fixes https://github.com/dotnet/runtime/issues/52646

### CODE REVIEW

**Comment by krwq** [comment]
@kronic are there any functional differences between those two calls? If yes please update the issue with all info, if not please update the issue we can easily tell the code is the same. This is needed for reference in case we get a regression or if this is intentional break we can easier point out what's the culprit or understand why this needs to happen.

**Comment by kronic** [comment]
@krwq Yes, there are differences in return types XmlAtomicValue, XPathItem, XmlAtomicValue[], XPathItem[], XPathNavigator[]

**Comment by krwq** [comment]
@kronic can you give more specific examples of differences? i.e. input and API output

**Comment by kronic** [comment]
@krwq 
```csharp
using System;
using System.IO;
using System.Xml;
using System.Xml.Schema;

const string xml = "<test a=\"sdf\"/>";

try
{
	using var xmlReader = XmlReader.Create(new StringReader(xml));
	xmlReader.MoveToContent();
	xmlReader.MoveToAttribute(0);

	if (xmlReader.ReadContentAs(typeof(XmlAtomicValue), null) is XmlAtomicValue xmlAtomic)
	{
		Console.WriteLine(xmlAtomic.Value);
	}
}
catch (Exception ex)
{
	Console.WriteLine(ex);
}

try
{
	using var xmlReader = XmlReader.Create(new StringReader(xml), new XmlReaderSettings {Async = true});
	await xmlReader.MoveToContentAsync();
	xmlReader.MoveToAttribute(0);

	if (await xmlReader.ReadContentAsAsync(typeof(XmlAtomicValue), null) is XmlAtomicValue xmlAtomic)
	{
		Console.WriteLine(xmlAtomic.Value);
	}
}
catch (Exception ex)
{
	Console.WriteLine(ex);
}
```

**Comment by kronic** [comment]
https://github.com/dotnet/runtime/issues/52646

**Comment by kronic** [comment]
@krwq ping

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs [hunk 1]
```diff
@@ -534,7 +534,7 @@ public virtual object ReadElementContentAs(Type returnType, IXmlNamespaceResolve
                 return value;
             }
 
-            return returnType == typeof(string) ? string.Empty : XmlUntypedStringConverter.Instance.FromString(string.Empty, returnType, namespaceResolver);
+            return returnType == typeof(string) ? string.Empty : XmlUntypedConverter.Untyped.ChangeType(string.Empty, returnType, namespaceResolver);
```
Comment (krwq) [inline src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: I believe the new implementation will not work with list types correctly. I think the correct solution might be just changing `Untyped` into `UntypedList` but you will need to add tests which cover this.

**Comment by krwq** [comment]
@kronic sorry for late re-review. You will need to add more tests related to list types as I think there are some differences in the implementation (by default XmlUntypedStringConverter.Instance handles lists while XmlUntypedConverter.Untyped doesn't). I think the solution is to change to XmlUntypedConverter.UntypedList but you will need to add tests to confirm everything work. Make sure to analyze both implementations and ensure there are no differences in behavior.

**Comment by jeffhandley** [comment]
@kronic I'm sorry we had a long delay on reviewing this. We will be ready to re-review if you want to address the feedback @krwq left.

Assigning to @krwq to track and follow-up as needed.

**Comment by krwq** [comment]
*no merge* until RC1 snap

**Comment by kronic** [comment]
@krwq done

File: src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs (hunk 1 above)
Comment (kronic) [reply src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReader.cs:537]: https://github.com/dotnet/runtime/blob/aa2af309a12d31dd68de9dc3646d77b6f25d9d40/src/libraries/System.Private.Xml/src/System/Xml/Core/XmlReaderAsync.cs#L61

File: src/libraries/System.Private.Xml/tests/XmlReader/ReadContentAs/ReadAsArrayTests.cs [hunk 2]
```diff
@@ -14,87 +16,611 @@ public static void DeserializationOfTypedArraysByXmlReader1()
...
             reader.Read();
             var values = (int[])reader.ReadContentAs(typeof(int[]), null);
-
             Assert.Equal(4, values.Length);
             Assert.Equal(1, values[0]);
             Assert.Equal(2, values[1]);
             Assert.Equal(3, values[2]);
             Assert.Equal(4, values[3]);
         }
+        [Fact]
```
Comment (krwq) [inline src/libraries/System.Private.Xml/tests/XmlReader/ReadContentAs/ReadAsArrayTests.cs:25]: nit: new line between `}` and `[Fact]` - here and everywhere

**Comment by krwq** [comment]
LGTM assuming no failures in outerloop tests - we should watch closely for any potential regressions caused by this PR since there might be some obscure cases which we might hit

**Review by krwq — APPROVED** [review]

**Pull request merged**